File: data-structure-algorithms/arrays/53_Maximum_Subarray.cpp
#include<iostream>
#include<vector>
using namespace std;


int maxSubArray(vector<int> nums)  {

    int max=INT_MIN;
    int maxi = 0;

    for (int i=0;i<nums.size();i++){
        maxi+=nums[i];

        if(maxi>max){
            max = maxi;
        }

        if(maxi <0){
            maxi = 0;
        }
    }

    return max;
}

int main(){

    vector<int> v = {-2, 1, -3, 4, -1, 2, 1, -5, 4};

    cout << maxSubArray(v);

    return 0;
}

File: data-structure-algorithms/arrays/53_maximum_subarray.go
package main

import "math"

func maxSubArray(nums []int) int {

	maxi := 0
	max := math.MinInt

	for i := 0; i < len(nums); i++ {
		maxi += nums[i]

		if maxi > max {
			max = maxi
		}
		if maxi < 0 {
			maxi = 0
		}
	}
	return max
}

func main() {

	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	result := maxSubArray(nums)

	println(result)
}
